Part 3 working with raster data (#3)
carpentries-incubator/r-geospatial-urban: episodes/16-raster-calculations.Rmd

---
title: 'Raster Calculations'
teaching: 25
exercises: 5
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
	echo = TRUE,
	message = FALSE,
	warning = FALSE
)

library(tidyverse)
library(terra)

DSM_TUD <- rast("data/tud-dsm-5m.tif")
DSM_TUD_df <- as.data.frame(DSM_TUD, xy = TRUE)

DTM_TUD <- rast("data/tud-dtm-5m.tif")
DTM_TUD_df <- as.data.frame(DTM_TUD, xy = TRUE)
```

:::::::::::::::::::::::::::::::::::::: questions 

- How do I subtract one raster from another and extract pixel values for defined locations?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Perform a subtraction between two rasters using raster math.
- Export raster data as a GeoTIFF file.

::::::::::::::::::::::::::::::::::::::::::::::::

::: prereq

# Things you'll need to complete this episode

> See the [lesson homepage]() for detailed information about the software, data, and other prerequisites you will need to work through the examples in this episode.

Author: ```suggestion
See the [lesson homepage](../index.md) for detailed information about the software, data, and other prerequisites you will need to work through the examples in this episode.
```

<!-- This lesson uses the `terra` package in particular. If you have not installed it yet, do so by running `install.packages("terra")` before loading it with `library(terra)`. -->

:::

> 

Author: I added an example and removed the last sentence (a stray sentence from the original lesson). See below.

> We often want to combine values of and perform calculations on rasters to create a new output raster. This episode covers how to subtract one raster from another using basic raster math and the `lapp()` function. It also covers how to extract pixel values from a set of locations - for example a buffer region around plot locations at a field site.

Author: ```suggestion
We often want to combine values of and perform calculations on rasters to create a new output raster. This episode covers how to subtract one raster from another using basic raster math and the `lapp()` function. For instance, this operation is useful to obtain building and tree heights by subtracting a raster that contains bare terrain (DTM) from a raster containing the terrain with heights of buildings and trees included (DSM). 
```

## Raster calculations in R

We often want to perform calculations on two or more rasters to create a new output raster. For example, if we are interested in mapping the heights of trees and buildings across an entire field site, we might want to calculate the difference between the Digital Surface Model (DSM, tops of trees and buildings) and the Digital Terrain Model (DTM, ground level). The resulting dataset is referred to as a Canopy Height Model (CHM) and represents the actual height of trees, buildings, etc. with the influence of ground elevation removed.

> ![Source: National Ecological Observatory Network (NEON).](https://datacarpentry.org/r-raster-vector-geospatial/fig/dc-spatial-raster/lidarTree-height.png)

Author: Sorry, I don't understand the comment.

> ![Source: National Ecological Observatory Network (NEON).](https://datacarpentry.org/r-raster-vector-geospatial/fig/dc-spatial-raster/lidarTree-height.png)

Reviewer: Hello @cforgaci , My comment is about this figure in the lesson:

![](https://datacarpentry.org/r-raster-vector-geospatial/fig/dc-spatial-raster/lidarTree-height.png)

The top-left figure is similar (almost the same) to the one on the bottom (CHM), and that is confusing by looking into the figure only. I understand the explanation on the text but it is not evident in the figure.

> ![Source: National Ecological Observatory Network (NEON).](https://datacarpentry.org/r-raster-vector-geospatial/fig/dc-spatial-raster/lidarTree-height.png)

Author: Thanks for pointing this out. You're right. It does remove the elevation in the latter, but it is not clear in this figure. I'll see if I can find an alternative, maybe including buildings too to fit our lesson better.

## Load the Data

For this episode, we will use the DTM and DSM data which we already have loaded from previous episodes.

We use the `describe()` function to view information about the DTM and DSM data files. 
```{r}
describe("data/tud-dtm-5m.tif")
describe("data/tud-dsm-5m.tif")
```

We’ve already loaded and worked with these two data files in earlier episodes. Let’s plot them each once more to remind ourselves what this data looks like. First we’ll plot the DTM elevation data:
```{r}
 ggplot() +
      geom_raster(data = DTM_TUD_df , 
              aes(x = x, y = y, fill = `tud-dtm-5m`)) +
     scale_fill_gradientn(name = "Elevation", colors = terrain.colors(10)) + 
     coord_quickmap()
```

And then the DSM elevation data:
```{r}
 ggplot() +
      geom_raster(data = DSM_TUD_df , 
              aes(x = x, y = y, fill = `tud-dsm-5m`)) +
     scale_fill_gradientn(name = "Elevation", colors = terrain.colors(10)) + 
     coord_quickmap()
```

> ## Raster math and Canopy Height Models

Author: I added this to the first sentence. I kept the title for consistency with the original lesson.

> We can perform raster calculations by subtracting (or adding, multiplying, etc) two rasters. In the geospatial world, we call this “raster math”.

Author: ```suggestion
We can perform raster calculations by subtracting (or adding, multiplying, etc) two rasters. In the geospatial world, we call this “raster math” or "map algebra".
```
